src/lib/utils/mem_ops.h

```diff
@@ -110,39 +157,114 @@ inline constexpr void copy_mem(T* out, const T* in, size_t n)
    }
 }
 
+/**
+* Copy memory
+* @param out the destination array
+* @param in the source array
+*/
+template <ranges::contiguous_output_range OutR, ranges::contiguous_range InR>
+inline constexpr void copy_mem(OutR&& out, InR&& in)
+   requires std::is_same_v<std::ranges::range_value_t<OutR>, std::ranges::range_value_t<InR>> &&
+            std::is_trivially_copyable_v<std::ranges::range_value_t<InR>>
+{
+   ranges::assert_equal_byte_lengths(out, in);
+   if(ranges::size_bytes(out) > 0) {
+      std::memmove(std::ranges::data(out), std::ranges::data(in), ranges::size_bytes(out));
+   }
+}
+
+/**
+ * Copy a rande of a trivially copyable type into another range of trivially
+ * copyable type of matching byte length.
+ */
+template <ranges::contiguous_output_range ToR, ranges::contiguous_range FromR>
+inline constexpr void typecast_copy(ToR&& out, FromR&& in)
+   requires std::is_trivially_copyable_v<std::ranges::range_value_t<FromR>> &&
+            std::is_trivially_copyable_v<std::ranges::range_value_t<ToR>>
```

Comment:
This is very closely related to C++20's [`std::bit_cast`](https://en.cppreference.com/w/cpp/numeric/bit_cast), right? I didn't find a way to use `std::bit_cast` on arbitrary ranges, though. Probably that's by design, as it is supposed to work closely with the type system.

But given that Botan's `typecast_copy` and STL's `bit_cast` are so closely related: Perhaps we should adopt the naming here and deprecate `typecast_copy` in the public API. For instance, we could call these functions `bitcast_copy`.